File: MarDaBE/src/main/java/com/wbill/home/service/BillingMeterRentService.java
package com.wbill.home.service;

import com.wbill.home.model.BillingMeterRent;
import com.wbill.home.model.BillingCustomerType;
import com.wbill.home.model.BillingMeterSize;
import com.wbill.home.repository.BillingMeterRentRepository;
import com.wbill.home.repository.BillingCustomerTypeRepository;
import com.wbill.home.repository.BillingMeterSizeRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class BillingMeterRentService {

    @Autowired
    private BillingMeterRentRepository meterRentRepository;
    
    @Autowired
    private BillingCustomerTypeRepository customerTypeRepository;
    
    @Autowired
    private BillingMeterSizeRepository meterSizeRepository;

    // Get all meter rents (both active and deleted for frontend categorization)
    public List<BillingMeterRent> getAllMeterRents() {
        try {
            // First attempt: get all meter rents
            List<BillingMeterRent> allMeterRents = meterRentRepository.findAllMeterRents();
            return allMeterRents;
        } catch (Exception e) {
            System.err.println("Error in getAllMeterRents: " + e.getMessage());
            // Fallback: get active meter rents only
            try {
                List<BillingMeterRent> activeMeterRents = meterRentRepository.findAllActive();
                return activeMeterRents;
            } catch (Exception e2) {
                System.err.println("Error in fallback getAllMeterRents: " + e2.getMessage());
                // Final fallback: return all from findAll
                return meterRentRepository.findAll();
            }
        }
    }

    // Get meter rents by status with pagination
    public Page<BillingMeterRent> getMeterRentsByStatus(String status, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("rentBirr").ascending());
        
        if ("ACTIVE".equalsIgnoreCase(status)) {
            return meterRentRepository.findAllActiveWithPagination(pageable);
        } else if ("DELETED".equalsIgnoreCase(status)) {
            return meterRentRepository.findAllDeleted(pageable);
        } else {
            return meterRentRepository.findByStatus(status.toLowerCase(), pageable);
        }
    }

    // Get meter rent by ID
    public Optional<BillingMeterRent> getMeterRentById(Integer id) {
        return meterRentRepository.findById(id);
    }

    // Create new meter rent
    public BillingMeterRent createMeterRent(BillingMeterRent meterRent) {
        // Validation
        validateMeterRent(meterRent, null);

        // Set default values
        meterRent.setStatus("active");

        return meterRentRepository.save(meterRent);
    }

    // Update existing meter rent
    public BillingMeterRent updateMeterRent(Integer id, BillingMeterRent updatedMeterRent) {
        Optional<BillingMeterRent> existingMeterRentOpt = meterRentRepository.findById(id);
        
        if (!existingMeterRentOpt.isPresent()) {
            throw new RuntimeException("Meter Rent not found with id: " + id);
        }

        BillingMeterRent existingMeterRent = existingMeterRentOpt.get();
        
        // Validation - skip duplicate validation during updates
        validateMeterRent(updatedMeterRent, id);

        // Update fields
        existingMeterRent.setBillingCustomerType(updatedMeterRent.getBillingCustomerType());
        existingMeterRent.setBillingMeterSize(updatedMeterRent.getBillingMeterSize());
        existingMeterRent.setRentBirr(updatedMeterRent.getRentBirr());

        return meterRentRepository.save(existingMeterRent);
    }

    // Soft delete (deactivate) meter rent
    public BillingMeterRent deactivateMeterRent(Integer id, String remark) {
        Optional<BillingMeterRent> meterRentOpt = meterRentRepository.findById(id);
        
        if (!meterRentOpt.isPresent()) {
            throw new RuntimeException("Meter Rent not found with id: " + id);
        }

        BillingMeterRent meterRent = meterRentOpt.get();
        meterRent.setStatus("deleted");

        return meterRentRepository.save(meterRent);
    }

    // Activate meter rent
    public BillingMeterRent activateMeterRent(Integer id) {
        Optional<BillingMeterRent> meterRentOpt = meterRentRepository.findById(id);
        
        if (!meterRentOpt.isPresent()) {
            throw new RuntimeException("Meter Rent not found with id: " + id);
        }

        BillingMeterRent meterRent = meterRentOpt.get();
        meterRent.setStatus("active");

        return meterRentRepository.save(meterRent);
    }

    // Helper method for validation
    private void validateMeterRent(BillingMeterRent meterRent, Integer excludeId) {
        if (meterRent.getBillingCustomerType() == null) {
            throw new IllegalArgumentException("Customer type is required");
        }

        if (meterRent.getBillingMeterSize() == null) {
            throw new IllegalArgumentException("Meter size is required");
        }

        if (meterRent.getRentBirr() < 0) {
            throw new IllegalArgumentException("Rent amount must be a positive number");
        }

        // Validate that customer type exists and is active
        Optional<BillingCustomerType> customerTypeOpt = customerTypeRepository.findById(meterRent.getBillingCustomerType().getId());
        if (!customerTypeOpt.isPresent() || !"active".equals(customerTypeOpt.get().getDeleted())) {
            throw new IllegalArgumentException("Invalid or inactive customer type");
        }

        // Validate that meter size exists and is active
        Optional<BillingMeterSize> meterSizeOpt = meterSizeRepository.findById(meterRent.getBillingMeterSize().getId());
        if (!meterSizeOpt.isPresent() || !"active".equals(meterSizeOpt.get().getDeleted())) {
            throw new IllegalArgumentException("Invalid or inactive meter size");
        }

        // Only check for duplicates during creation (excludeId is null)
        if (excludeId == null) {
            if (meterRentRepository.existsByBillingCustomerTypeAndBillingMeterSize(
                    meterRent.getBillingCustomerType(), meterRent.getBillingMeterSize())) {
                throw new IllegalArgumentException("Meter rent for this customer type and meter size combination already exists");
            }
        }
        // Skip duplicate validation during updates
    }

    // Get meter rent statistics
    public MeterRentStatistics getMeterRentStatistics() {
        Long activeCount = meterRentRepository.countActiveMeterRents();
        Long deletedCount = meterRentRepository.countDeletedMeterRents();
        
        return new MeterRentStatistics(
            activeCount != null ? activeCount : 0L,
            deletedCount != null ? deletedCount : 0L
        );
    }

    // Inner class for statistics
    public static class MeterRentStatistics {
        private Long activeMeterRents;
        private Long deletedMeterRents;

        public MeterRentStatistics(Long activeMeterRents, Long deletedMeterRents) {
            this.activeMeterRents = activeMeterRents;
            this.deletedMeterRents = deletedMeterRents;
        }

        public Long getActiveMeterRents() {
            return activeMeterRents;
        }

        public void setActiveMeterRents(Long activeMeterRents) {
            this.activeMeterRents = activeMeterRents;
        }

        public Long getDeletedMeterRents() {
            return deletedMeterRents;
        }

        public void setDeletedMeterRents(Long deletedMeterRents) {
            this.deletedMeterRents = deletedMeterRents;
        }
    }
}

File: MarDaFE/app/lib/billingMeterRentService.js
"use client";
import authHeader from "./authHeader/authhheader";
import getAccesToken from "./getToken";
import axios from "axios";
import { baseURL } from "./httpCommon/http-common";

const baseUrl = new baseURL();
const commonUrl = baseUrl.getUrl();


export class BillingMeterRentService {
  // Get all meter rents
  async getAllMeterRents() {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.get(`${commonUrl}billing-meter-rents/all`, {
        headers: authHeader(token),
      });
      return res.data;
    } catch (error) {
      console.error("Error fetching all meter rents:", error);
      throw error;
    }
  }

  // Get meter rents by status with pagination
  async getMeterRentsByStatus(status, page = 0, size = 10) {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.get(
        `${commonUrl}billing-meter-rents/status/${status}?page=${page}&size=${size}`,
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error fetching meter rents by status:", error);
      throw error;
    }
  }

  // Get meter rent by ID
  async getMeterRentById(id) {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.get(`${commonUrl}billing-meter-rents/${id}`, {
        headers: authHeader(token),
      });
      return res.data;
    } catch (error) {
      console.error("Error fetching meter rent by ID:", error);
      throw error;
    }
  }

  // Create new meter rent
  async createMeterRent(meterRentData) {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.post(
        `${commonUrl}billing-meter-rents`,
        meterRentData,
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error creating meter rent:", error);
      throw error;
    }
  }

  // Update existing meter rent
  async updateMeterRent(id, meterRentData) {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.put(
        `${commonUrl}billing-meter-rents/${id}`,
        meterRentData,
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error updating meter rent:", error);
      throw error;
    }
  }

  // Deactivate meter rent
  async deactivateMeterRent(payload) {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.post(
        `${commonUrl}billing-meter-rents/${payload.id}/deactivate`,
        { remark: payload.remark || "" },
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error deactivating meter rent:", error);
      throw error;
    }
  }

  // Activate meter rent
  async activateMeterRent(id) {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.post(
        `${commonUrl}billing-meter-rents/${id}/activate`,
        {},
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error activating meter rent:", error);
      throw error;
    }
  }

  // Get meter rent statistics
  async getMeterRentStatistics() {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.get(
        `${commonUrl}billing-meter-rents/statistics`,
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error fetching meter rent statistics:", error);
      throw error;
    }
  }

  // Get active customer types for dropdown
  async getActiveCustomerTypes() {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.get(
        `${commonUrl}billing-meter-rents/customer-types`,
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error fetching active customer types:", error);
      throw error;
    }
  }

  // Get active meter sizes for dropdown
  async getActiveMeterSizes() {
    try {
      const token = localStorage.getItem("token");
      const res = await axios.get(
        `${commonUrl}billing-meter-rents/meter-sizes`,
        {
          headers: authHeader(token),
        }
      );
      return res.data;
    } catch (error) {
      console.error("Error fetching active meter sizes:", error);
      throw error;
    }
  }
}
